spices/SPICE-0003-windows-safe-paths.adoc

```diff
@@ -0,0 +1,137 @@
+= Windows Safe Paths
+
+* Proposal: xref:SPICE-0003-windows-safe-paths.adoc[SPICE-0003]
+* Author: link:https://github.com/bioball[Dan Chao]
+* Status: TBD
+* Implemented in: TBD
+* Category: Language
+
+== Introduction
+
+In order to support running Pkl on Windows, filepaths are encoded to be compatible on Windows operating systems.
+
+== Motivation
+
+Currently, Pkl is only supported for macOS and Linux.
+On these operating systems, the only character reserved in filepaths is `/` (the directory separator).
+
+On Windows, the characters `<`, `>`, `:`, `"`, `\`, `|`, `?`, and `*` are also reserved.
+
+In situations where filepaths are written to disk, operations to write filepaths that contain these characters will either fail or have otherwise unexpected behavior.
+
+== Proposed Solution
+
+In order to be compatible with Windows, filepaths will be encoded to be compatible with Windows.
+
+These are the situations where Pkl will write encoded paths to disk:
+
+1. Writing packages to the cache directory, when using `--cache-dir` or when using the default cache dir.
+2. Creating documentation via `pkldoc`, or the `pkl-doc` Java library.
+3. Generating Kotlin code through `pkl-codegen-kotlin`.
+
+NOTE: Java identifiers cannot use any of the reserved Windows characters, so we do not need to change the behavior of `pkl-codegen-java`.
+
+=== Files written from `pkl eval` are not escaped
+
+Pkl will _not_ write encoded filepaths during regular evaluation.
+When executing `pkl eval -o "foo:bar" myModule.pkl` on Windows, Pkl will attempt to write a filename called `foo:bar`, the I/O error is bubbled up to the user.
+
+Additionally, the paths specified within `output.files` when using link:https://pkl-lang.org/main/current/language-reference/index.html#multiple-file-output[multiple-file output] will be written verbatim without escaping, and any I/O errors are bubbled up to the user.
+
+These paths are not encoded because:
+
+1. They are user-specified. +
+It can be a confusing experience if `pkl eval -o foo:bar` produced a _different_ filename on operating systems where this is supported.
+It can also be a confusing experience if Pkl behaved differently on different operating systems.
+2. It preserves compatiblity with current Pkl.
+
+== Detailed design
+
+=== Filename encoding
+
+Characters that aren't safe on Windows are encoded by enclosing their hexadecimal character code in parentheses.
+
+The open parenthesis literal `(` is encoded as `((`.
```

Comment:
I think we should also encode `)` as `))` for consistency.
Also, why are there illegal cases if they can be properly encoded? `foo(az)bar` -> `foo((az))bar`

Reply:
That example shows that `foo(az)bar` is an illegal encoding, because `az` is not a hex code.

Reply:
Summary from offline discussion: we will not escape the ending parenthesis (we won't encode if we don't need to).

Regex also works this way; the closing `]` in a character class doesn't need to be escaped.